### PR TITLE
[Dev] Fix DSA sparse attention backward empty rows

### PR DESCRIPTION
**Pull request by buptzyb**
- [ ] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?

Main counterpart: #6939.

Depends on #6206. This PR is intentionally stacked on the upstream
`pull-request/6206` branch; review only the two commits above that branch. If
#6206 is updated, this branch will be rebased onto its new head. After #6206
merges, this PR will be rebased onto and retargeted to `dev`.

Keep ordinary DSA sparse-attention backward shapes static when top-k contains
empty query rows, removing its dynamic `nonzero`/compact/gather/scatter path.

The generic path replaces each empty row with a safe one-tile cuDNN invocation:
its first index is zero, its effective top-k length is one, and its `dO`/`LSE`
are zeroed before backward. The returned `dQ` is explicitly zeroed for those
rows, so the safe tile contributes neither query nor KV gradients. A conservative
host-side certificate retains the existing allocation-free fast path when all
rows are provably nonempty.

This covers both the combined fused path and the split path used by IndexShare
and MTP sharing. It does not change CSA, THD/CP layout, MTP sharing semantics,
configuration, or the full-iteration CUDA Graph validator.

## Issue tracking

Linked issue: N/A

### Pre-checks

- [x] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [x] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [x] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Verification notes

- Unit/native parity: full and split paths; nonempty, mixed-empty, and all-empty
  rows; BF16/FP32; metadata immutability; zero-size rejection; batch isolation.
- H100 real cuDNN: split forward/backward and changing-pattern CUDA Graph replay
  passed. The combined cuDNN IndexerTopK forward is SM100-only and was covered on
  GB200.
- GB200 real cuDNN: split and combined eager/native parity and fixed-shape graph
  replay passed; no stale mask or replay allocation growth.
- 16-GPU GB200 GLM proxy ABBA: short THD 4K effective CP1 improved 2.38%; long
  THD 64K dynamic CP16 improved 3.19%; peak memory was unchanged in both cells.
- Nsight Systems 2026.4.1: candidate DSA backward traces contain no target
  `nonzero`, `index_select`, `index_copy`, or dynamic gather/scatter operations.

### CODE REVIEW

File: megatron/core/transformer/experimental_attention_variant/dsa_cudnn_kernels.py [hunk 1]
```diff
@@ -2620,11 +2611,12 @@ def run_fused_absorbed_sparse_attention(
     softmax_scale: float,
     v_channels: int,
     topk_length: Optional[Tensor] = None,
+    all_topk_rows_nonempty: bool = False,
 ) -> Optional[Tensor]:
     """Run cuDNN/FlashMLA sparse attention using externally supplied top-k indices."""
     if query.ndim != 4 or key.ndim != 4 or topk_indices.ndim != 3:
         return None
-    if key.size(2) != 1 or v_channels <= 0:
+    if key.size(0) == 0 or key.size(2) != 1 or topk_indices.size(2) == 0 or v_channels <= 0:
```
[the remark this answers is not in the document]
Comment (buptzyb) [reply megatron/core/transformer/experimental_attention_variant/dsa_cudnn_kernels.py:2947]: Addressed in f61604442. Both entry points now explain that nonpositive/zero-width top-k or empty KV cannot provide a valid slot for the fixed-shape backward safe tile, so the fused path must decline rather than index an empty dimension. This records the correctness dependency separately from the surrounding shape support checks.

File: megatron/core/transformer/experimental_attention_variant/dsa_cudnn_kernels.py [hunk 2]
```diff
@@ -2228,6 +2203,23 @@ def forward(
...
+                or (
+                    use_local_indexer_varlen
+                    and varlen_starts is not None
+                    and varlen_ends is not None
+                    and key_positions is None
+                )
+            )
+        )
+        if not all_topk_rows_nonempty:
+            global_idxs[:, 0].masked_fill_(topk_length_flat <= 0, 0)
```
[the remark this answers is not in the document]
Comment (buptzyb) [reply megatron/core/transformer/experimental_attention_variant/dsa_cudnn_kernels.py:2542]: The literal global index 0 is intentional after making the dummy tile numerically inert. Computing a same-batch dummy would add an arange/where allocation and kernel on this backward-critical path. Under the pinned FlashMLA contract O=0, and the wrapper now forces dO=0 and LSE=+inf, so D=0, P=0, dS=0, and dV=0. The tile may read batch 0 but cannot contribute a cross-batch gradient. Commit f61604442 adds this invariant beside both assignments and renames the batched regression to test_cudnn_attention_backward_dummy_rows_add_no_cross_batch_gradient so the test states what it actually proves.

File: tests/unit_tests/transformer/experimental_attention_variant/test_dsa_native_parity.py [hunk 3]
```diff
@@ -1903,75 +1907,195 @@ def sparse_attention_backward_wrapper(
...
-            torch.tensor([[0, 2]], dtype=torch.int32),
-            None,
-        )
+        return make_topk(4), torch.tensor([row_lengths], dtype=torch.int32), None
 
     def fake_flash_mla(q, kv, topk_idxs, softmax_scale, d_v, attn_sink, topk_length):
         seen["fwd_topk"] = topk_idxs.detach().clone()
         seen["fwd_topk_length"] = topk_length.detach().clone()
-        return torch.zeros((2, 1, d_v), dtype=q.dtype), torch.zeros((2, 1), dtype=torch.float32)
+        return torch.zeros((2, 1, d_v), dtype=q.dtype), torch.tensor([[5.0], [6.0]])
```
[the remark this answers is not in the document]
Comment (buptzyb) [reply tests/unit_tests/transformer/experimental_attention_variant/test_dsa_native_parity.py]: Updated in f61604442. The fake now matches the pinned FlashMLA empty-row contract: O=0 and LSE=+inf. The regression asserts that backward preserves +inf while sanitizing dO for full-fusion and split-attention, mixed and all-empty rows, in fp32 and bf16. We intentionally do not inject NaN into O because that state is excluded by the supported FlashMLA contract verified in the kernel sources; the separate b>1 dummy-row regression proves no cross-batch gradient is added even with NaN grad_output. The all-nonempty full-fusion certificate remains covered by test_cudnn_full_fusion_uses_nonempty_local_varlen_certificate.

File: megatron/core/transformer/experimental_attention_variant/dsa_cudnn_kernels.py [hunk 4]
```diff
@@ -2228,6 +2203,23 @@ def forward(
...
+            and (
+                (varlen_starts is None and varlen_ends is None and key_positions is None)
+                or (
+                    use_local_indexer_varlen
+                    and varlen_starts is not None
+                    and varlen_ends is not None
+                    and key_positions is None
+                )
+            )
+        )
```
[the remark this answers is not in the document]
Comment (buptzyb) [reply megatron/core/transformer/experimental_attention_variant/dsa_cudnn_kernels.py]: Addressed in f61604442. The full-fusion predicate now explicitly cross-references dsa._can_prove_all_topk_rows_nonempty and documents why it is intentionally backend-normalized: plain-causal bounds have already been normalized to None, while the attention-mask and computes-topk gates are implied by dispatch. This keeps the two stages distinct without introducing a helper whose inputs would obscure the normalization boundary.

File: megatron/core/transformer/experimental_attention_variant/dsa_kernels.py [hunk 5]
```diff
@@ -89,6 +91,29 @@ def _resolve_fused_hook(config: TransformerConfig, hook_name: str):
...
+        return None
+    return frozenset(signature.parameters)
+
+
+def _hook_kwargs_accepting(fn, **candidate_kwargs):
+    """Keep only optional internal kwargs accepted by a backend hook."""
+    parameter_names = _hook_parameter_names(fn)
+    if parameter_names is None:
+        return candidate_kwargs
+    return {name: value for name, value in candidate_kwargs.items() if name in parameter_names}
```
[the remark this answers is not in the document]
Comment (buptzyb) [reply megatron/core/transformer/experimental_attention_variant/dsa_kernels.py]: Addressed in f61604442. The signature cache is now bounded at maxsize=128, avoiding unbounded strong retention of dynamically monkeypatched callables. The routing helper docstring now states the compatibility invariant explicitly: candidate kwargs must be safe to omit for legacy hooks, and hooks accepting all_topk_rows_nonempty must conservatively default it to False.

File: megatron/core/transformer/experimental_attention_variant/dsa_cudnn_kernels.py [hunk 6]
```diff
@@ -2071,56 +2073,29 @@ def _run_sparse_attention_backward(
...
-    bwd_out_flat = torch.cat((bwd_out_flat, torch.zeros_like(bwd_out_flat[:1])), dim=0)
-    bwd_dO_flat = torch.cat((bwd_dO_flat, torch.zeros_like(bwd_dO_flat[:1])), dim=0)
-    bwd_lse = torch.cat((bwd_lse, torch.zeros_like(bwd_lse[:1])), dim=0)
-    global_idxs = torch.cat((global_idxs, torch.zeros_like(global_idxs[:1])), dim=0)
-    topk_length = torch.cat((topk_length, torch.ones_like(topk_length[:1])), dim=0)
+    empty_rows = None
+    if not all_topk_rows_nonempty:
+        empty_rows = topk_length <= 0
+        topk_length = topk_length.masked_fill(empty_rows, 1)
+        bwd_dO_flat = bwd_dO_flat.masked_fill(empty_rows[:, None, None], 0)
```
[the remark this answers is not in the document]
Comment (buptzyb) [reply megatron/core/transformer/experimental_attention_variant/dsa_cudnn_kernels.py:2394]: Thanks for flagging this. We verified the contract at the exact FlashMLA revision pinned by this PR (b7643bd54521f563b839b98289b5cd048c062ba2). For an empty row, SM90 sets the output scale to zero and stores LSE=+inf ([output path](https://github.com/deepseek-ai/FlashMLA/blob/b7643bd54521f563b839b98289b5cd048c062ba2/csrc/sm90/prefill/sparse/phase1.cuh#L206-L212), [LSE path](https://github.com/deepseek-ai/FlashMLA/blob/b7643bd54521f563b839b98289b5cd048c062ba2/csrc/sm90/prefill/sparse/phase1.cuh#L494-L500)); the SM100 head64/head128 paths likewise zero O when there are no valid indices and convert empty LSE to +inf ([head64](https://github.com/deepseek-ai/FlashMLA/blob/b7643bd54521f563b839b98289b5cd048c062ba2/csrc/sm100/prefill/sparse/fwd/head64/phase1.cuh#L274-L323), [head128](https://github.com/deepseek-ai/FlashMLA/blob/b7643bd54521f563b839b98289b5cd048c062ba2/csrc/sm100/prefill/sparse/fwd/head128/phase1.cuh#L304-L336)). Thus, after this wrapper zeroes dO, D = rowsum(dO * O) is exactly zero; an empty-row O cannot be non-finite under the supported backend contract. Re-masking O would add another full-tensor write/allocation on the backward path. Commit f61604442 documents this dependency and preserves LSE=+inf; the regression fake intentionally follows the pinned FlashMLA contract.

File: megatron/core/transformer/experimental_attention_variant/dsa_cudnn_kernels.py [hunk 7]
```diff
@@ -2071,56 +2073,29 @@ def _run_sparse_attention_backward(
...
+        bwd_lse = bwd_lse.masked_fill(empty_rows[:, None], 0)
```
[the remark this answers is not in the document]
Comment (buptzyb) [reply megatron/core/transformer/experimental_attention_variant/dsa_cudnn_kernels.py]: Fixed in f61604442. Empty rows are now promoted to one safe tile with dO=0 and LSE=+inf, so P = exp(S - LSE) is zero regardless of the recomputed logit magnitude. We use +inf rather than a large finite sentinel because +inf is the pinned FlashMLA empty-row representation and is handled by the cuDNN backward path. The full-fusion and split-attention regression cases now assert +inf for mixed and all-empty rows in both fp32 and bf16.

File: megatron/core/transformer/experimental_attention_variant/dsa.py [hunk 8]
```diff
@@ -193,6 +195,36 @@ def _run_sparse_attention(
...
+        or attention_mask is not None
+        or query_valid_rows is not None
+    ):
+        return False
+    return varlen_is_plain_causal or (
+        use_local_indexer_varlen
+        and varlen_starts is not None
+        and varlen_ends is not None
+        and key_positions is None
+    )
```
[the remark this answers is not in the document]
Comment (buptzyb) [reply megatron/core/transformer/experimental_attention_variant/dsa.py]: Thanks; we traced this through the in-tree packed layout and added both documentation and regression coverage in f61604442. The proposed empty-range derivation is reversed: for qpos >= cu_seqlens[-1], searchsorted returns the terminal sequence index, so start = cu_seqlens[-1] while end = qpos + 1, hence start < end. The certificate is provenance-based on the internal DSA packed self-attention layout: selected cumulative lengths (padded when available), synthetic tail positions, and reordered KV cover every physical query row. The new CPU regression uses two compact/padded sequences, between-sequence and tail padding, CP=2 across both ranks, and no real_token_mask_q; it verifies every physical query bound intersects KV and that the certificate is true. The docstring also states that tensor values are not rechecked here because doing so would introduce a device sync.

**Review by jingqiny-99 — APPROVED** [review]
LGTM
